### PR TITLE
[Tool] New tool to download android dependencies 

### PR DESCRIPTION
**Pull request by reidbaker**
This pr is pushed for high level feedback/conversation. I will add tests before serious review. 
should be read in conjuction with https://flutter-review.googlesource.com/c/recipes/+/46980

- Create new top level command to run flutter dependencies on changed packages
- when running android tests download dependencies before running tests

https://github.com/flutter/flutter/issues/120119

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [ ] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: .ci/targets/android_platform_tests.yaml [hunk 1]
```diff
@@ -7,6 +7,10 @@ tasks:
   - name: lint
     script: script/tool_runner.sh
     args: ["lint-android"]
+  - name: download android deps
+    script: script/tool_runner.sh
+    infra_step: true
```
Comment (stuartmorgan) [inline .ci/targets/android_platform_tests.yaml:12]: Nice! I didn't realize they could be intermixed with other steps; this is much simpler that what I thought we'd have to do.

File: script/tool/lib/src/fetch_gradle_deps_command.dart [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import 'common/gradle.dart';
+import 'common/package_looping_command.dart';
+import 'common/plugin_utils.dart';
+import 'common/process_runner.dart';
+import 'common/repository_package.dart';
+
+/// Run 'gradlew dependencies'.
+///
+/// See https://docs.gradle.org/6.4/userguide/core_dependency_management.html#sec:dependency-mgmt-in-gradle.
+class FetchGradleDeps extends PackageLoopingCommand {
```
Comment (stuartmorgan) [inline script/tool/lib/src/fetch_gradle_deps_command.dart:18]: Is there any way we can do a shared cache, or does Gradle caching only work at the project level?

File: script/tool/lib/src/fetch_gradle_deps_command.dart [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+      final String packageName = package.directory.basename;
+
+      // Only lint one build mode to avoid extra work.
+      // Only lint the plugin project itself, to avoid failing due to errors in
+      // dependencies.
+      //
+      // TODO(stuartmorgan): Consider adding an XML parser to read and summarize
+      // all results. Currently, only the first three errors will be shown
+      // inline, and the rest have to be checked via the CI-uploaded artifact.
```
Comment (stuartmorgan) [inline script/tool/lib/src/fetch_gradle_deps_command.dart:65]: I think you probably don't want this whole comment block ;)

**Review by stuartmorgan — COMMENTED** [review]
This looks great! If there's a way to share the cache for the whole test run rather than per-plugin that would be even better (or is Gradle actually doing that under the hood?), but either way this is a clear improvement, and should really help us quickly diagnose network and server flake.

**Comment by reidbaker** [comment]
> https://flutter-review.googlesource.com/c/recipes/+/46980

So our testing system is complex but gradle lays out the files it relies on [here](https://docs.gradle.org/current/userguide/directory_layout.html) by default all tests on a machine will share the gradle cache unless we configure it differently. Searching I was not able to find a server lifecycle doc to see how often those machines are wiped [1]. Searching for where we might configure gradle home I found a bunch of places where we set [GRADLE_USER_HOME](https://cs.opensource.google/search?q=GRADLE_USER_HOME%20-f:gradle-wrapper.properties&start=1) and all the the values we set are the same` [cache]/gradle` value. So those places should share a gradle directory which holds both dependencies and gradle wrapper version and cleans itself of unused versions every  7-30 days depending on the type of thing in the cache. 

[1] Moma search, exploring https://flutter.googlesource.com/infra/

File: script/tool/lib/src/fetch_gradle_deps_command.dart (hunk 2 above)
Comment (reidbaker) [reply script/tool/lib/src/fetch_gradle_deps_command.dart:18]: Gradle uses a shared cache as a default.

File: .ci/targets/android_platform_tests.yaml [hunk 4]
```diff
@@ -7,6 +7,10 @@ tasks:
   - name: lint
     script: script/tool_runner.sh
     args: ["lint-android"]
+  - name: download android deps
```
Comment (keyonghan) [inline .ci/targets/android_platform_tests.yaml:10]: Based on conversations from https://flutter-review.googlesource.com/c/recipes/+/46980, it is recommended to move the infra step to the very beginning.

**Comment by keyonghan** [comment]
> Searching for where we might configure gradle home I found a bunch of places where we set [GRADLE_USER_HOME](https://cs.opensource.google/search?q=GRADLE_USER_HOME%20-f:gradle-wrapper.properties&start=1) and all the the values we set are the same [cache]/gradle value. So those places should share a gradle directory which holds both dependencies and gradle wrapper version and cleans itself of unused versions every 7-30 days depending on the type of thing in the cache.

Do you have a build example which download gradle dependencies that I can take a look? `[cache]/gradle` relies on swarming named cache, and we should have a corresponding named cache entry defined so that different tests can share the cache on the same testbed. Based on existing infra config, I do not see any named cache pointing to `[cache]/gradle`.

File: script/tool/lib/src/fetch_deps_command.dart [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+// found in the LICENSE file.
+
+
+import 'common/core.dart';
+import 'common/gradle.dart';
+import 'common/package_looping_command.dart';
+import 'common/plugin_utils.dart';
+import 'common/repository_package.dart';
+
+/// Run 'gradlew dependencies' on android.
```
Comment (johnmccutchan) [inline script/tool/lib/src/fetch_deps_command.dart:12]: Update this comment now that this command is more generic.

**Review by johnmccutchan — APPROVED** [review]

**Comment by reidbaker** [comment]
> > Searching for where we might configure gradle home I found a bunch of places where we set [GRADLE_USER_HOME](https://cs.opensource.google/search?q=GRADLE_USER_HOME%20-f:gradle-wrapper.properties&start=1) and all the the values we set are the same [cache]/gradle value. So those places should share a gradle directory which holds both dependencies and gradle wrapper version and cleans itself of unused versions every 7-30 days depending on the type of thing in the cache.
> 
> Do you have a build example which downloads gradle dependencies that I can take a look? `[cache]/gradle` relies on swarming named cache, and we should have a corresponding named cache entry defined so that different tests can share the cache on the same testbed. Based on existing infra config, I do not see any named cache pointing to `[cache]/gradle`.

This is the only pr that should have the new step to download dependencies. That said all existing android test would have gradle dependencies because I think we include kotlin by default. 

Here is a recent build https://ci.chromium.org/ui/p/flutter/builders/try/Linux_android%20android_build_all_packages%20master/505/overview

I am not sure if I understand your comment about the existing infra config all I have found is in https://cs.opensource.google/search?q=GRADLE_USER_HOME%20-f:gradle-wrapper.properties&start=1.

**Comment by reidbaker** [comment]
Tests failing https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8775849073271137377/+/u/Run_package_tests/download_android_deps/stdout 

```
Unhandled exception:
Could not find an option named "packages-for-branch".
```

**Review by keyonghan — APPROVED** [review]
`android_platform_tests.yaml` LGTM

**Review by johnmccutchan — APPROVED** [review]

**Pull request merged**